SI-9546 Fix regression in rewrite of case apply to constructor call

retronym opened the pull request:
In SI-9425, I disabled the rewrite of `CaseClass.apply(x)` to
`new CaseClass(x)` if the constructor was was less accessible
than the apply method. This solved a problem with spurious
"constructor cannot be accessed" errors during refchecks for
case classes with non-public constructors.

However, for polymorphic case classes, refchecks was persistent,
and even after refusing to transform the `TypeApply` within:

```
CaseClass.apply[String]("")
```

It _would_ try again to transform the enclosing `Select`, a
code path only intended for monomorphic case classes. The tree has
a `PolyType`, which foiled the newly added accessibility check.

I've modified the guard to specifically exclude `PolyType` typed
trees (which only reach that code after we decline to transform
an enclosing `TypeApply`). I've also rewritten the check to use
`finalResultType` rather than `resultType`, which would have also
avoided the problem.

```
scala> case class CaseClass[A](a: A)
defined class CaseClass

scala> typeOf[CaseClass.type].decl(TermName("apply")).info
res0: $r.intp.global.Type = [A](a: A)CaseClass[A]

scala> res0.resultType
res1: $r.intp.global.Type = (a: A)CaseClass[A]

scala> res0.resultType.typeSymbol
res2: $r.intp.global.Symbol = <none>

scala> res0.finalResultType
res3: $r.intp.global.Type = CaseClass[A]

scala> res0.finalResultType.typeSymbol
res4: $r.intp.global.Symbol = class CaseClass
```

I've manually checked that the rewrite is still applied for a
generic case class with a public constructor.

```
% ./build-sbt/pack/bin/scalac -Xprint:refchecks test/files/run/t9546.scala 2>&1 | grep "def mkOpt"
      def mkOpt: foo.Opt[String] = Opt.apply[String]("");
/code/scala2 on ticket/9546*
% ./build-sbt/pack/bin/scalac -Xprint:refchecks test/files/run/t9546b.scala 2>&1 | grep "def mkOpt"
      def mkOpt: foo.Opt[String] = new foo.Opt[String]("");
```

If the case class was not also a value class, the bug still existed,
but never manifest as a crash. In `run/t9546c.scala`, the following
nonsensical tree was created in refchecks, but the problem was washed
away during erasure.

```
def mkOpt = new foo.Opt[A][String]("");
```

If `Opt` is a value class, `posterasure` needs to inspect and transform
this tree, and the crash ensued.

retronym commented:
IDE build failure: https://scala-ci.typesafe.com/job/scala-2.11.x-integrate-ide/2200/

retronym commented:
/rebuild

SethTisue commented:
I've opened https://github.com/scala/scala-dev/issues/96 on the integrate-ide regression

lrytz commented on src/compiler/scala/tools/nsc/typechecker/RefChecks.scala:
```diff
@@ -1520,11 +1520,12 @@ abstract class RefChecks extends InfoTransform with scala.reflect.internal.trans
       }
 
       val doTransform =
+        !tree.tpe.isInstanceOf[PolyType] && // SI-9546 If we've bailed out of the transform for a TypeApply, don't try again for the enclosed Select
```
i'd leave this line out given that the fix below is enough

retronym replied:
This one actually makes more sense as the primary check. Otherwise, in the case of a public constructor, we are relying on the details of the transform (e.g. that parents are transformed before children).

lrytz replied:
hmm, i don't see why this is, sorry..

retronym replied:
I'll expand the internal docs tomorrow to help explain this.

retronym committed:
Refactor transform of case apply in refchecks

I've identified a dead call to `transformCaseApply` that
seems to date back to Scala 2.6 vintages, in which case
factory methods were a fictional companion method, rather
than a real apply method in a companion module.

This commit adds an abort in that code path to aide code
review (if our test suite still passes, we know that I've
removed dead code, rather than silently changing behaviour.)
The following commit will remove it altogether

I then inlined a slightly clunky abstraction in the two
remaining calls to `transformCaseApply`. It was getting in the
way of a clean fix to SI-9546, the topic of the next commit.

retronym committed:
SI-9546 Fix regression in rewrite of case apply to constructor call

In SI-9425, I disabled the rewrite of `CaseClass.apply(x)` to
`new CaseClass(x)` if the constructor was was less accessible
than the apply method. This solved a problem with spurious
"constructor cannot be accessed" errors during refchecks for
case classes with non-public constructors.

However, for polymorphic case classes, refchecks was persistent,
and even after refusing to transform the `TypeApply` within:

    CaseClass.apply[String]("")

It *would* try again to transform the enclosing `Select`, a
code path only intended for monomorphic case classes. The tree has
a `PolyType`, which foiled the newly added accessibility check.

I've modified the call to `isSimpleCaseApply` from the transform
of `Select` nodes to exclude polymorphic apply's from being
considered twice.

retronym commented:
@lrytz Hopefully this commit progression will make a bit more sense.

SethTisue commented:
/synch

SethTisue commented:
/rebuild

SethTisue commented:
community build: https://scala-ci.typesafe.com/job/scala-2.11.x-integrate-community-build/242/consoleFull

SethTisue commented:
@lrytz if you could review this sometime today, that'd be good — I'd like to get the 2.11.8 release train rolling again

lrytz commented:
I think we still need the `finalResultType` in https://github.com/retronym/scala/commit/1a59c55ed5c6a0d53acd0259090cf57328748451#diff-e2cfc09e4d7c753232666c6e31bd1926R1525

Example:
```
case class A private (x: Int)
case class B private (x: Int)(y: Int)

class C {
  def f = A(1)
  def g = B(1)(2)
}
```

With 2.11.7:
```
➜  sandbox git:(pr5002) scalac11 -version
Scala compiler version 2.11.7 -- Copyright 2002-2013, LAMP/EPFL
➜  sandbox git:(pr5002) scalac11 Test.scala
Test.scala:5: error: constructor A in class A cannot be accessed in class C
  def f = A(1)
          ^
Test.scala:6: error: constructor B in class B cannot be accessed in class C
  def g = B(1)(2)
          ^
two errors found
``


➜  sandbox git:(pr5002) ✗ git diff | cat
➜  sandbox git:(pr5002) qsc -version
Scala compiler version 2.11.8-20160302-221155-fe5bd09861 -- Copyright 2002-2016, LAMP/EPFL
➜  sandbox git:(pr5002) qsc Test.scala
Test.scala:6: error: constructor B in class B cannot be accessed in class C
  def g = B(1)(2)
          ^
one error found
```

lrytz commented:
I think we still need the `finalResultType` in https://github.com/retronym/scala/commit/1a59c55ed5c6a0d53acd0259090cf57328748451#diff-e2cfc09e4d7c753232666c6e31bd1926R1525

Example:
```
case class A private (x: Int)
case class B private (x: Int)(y: Int)

class C {
  def f = A(1)
  def g = B(1)(2)
}
```

With 2.11.7:
```
➜  sandbox git:(pr5002) scalac11 -version
Scala compiler version 2.11.7 -- Copyright 2002-2013, LAMP/EPFL
➜  sandbox git:(pr5002) scalac11 Test.scala
Test.scala:5: error: constructor A in class A cannot be accessed in class C
  def f = A(1)
          ^
Test.scala:6: error: constructor B in class B cannot be accessed in class C
  def g = B(1)(2)
          ^
two errors found
```

With fe5bd09861994734bc394813d069ea40c89d39de

```
➜  sandbox git:(pr5002) qsc -version
Scala compiler version 2.11.8-20160302-221155-fe5bd09861 -- Copyright 2002-2016, LAMP/EPFL
➜  sandbox git:(pr5002) qsc Test.scala
Test.scala:6: error: constructor B in class B cannot be accessed in class C
  def g = B(1)(2)
          ^
one error found
```

With `finalResultType`
```
➜  sandbox git:(pr5002) ✗ git diff | cat
diff --git a/src/compiler/scala/tools/nsc/typechecker/RefChecks.scala b/src/compiler/scala/tools/nsc/typechecker/RefChecks.scala
index 517271e..c19e11b 100644
--- a/src/compiler/scala/tools/nsc/typechecker/RefChecks.scala
+++ b/src/compiler/scala/tools/nsc/typechecker/RefChecks.scala
@@ -1522,7 +1522,7 @@ abstract class RefChecks extends InfoTransform with scala.reflect.internal.trans
         sym.isCase &&
         sym.name == nme.apply &&
         isClassTypeAccessible(tree) &&
-        !tree.tpe.resultType.typeSymbol.primaryConstructor.isLessAccessibleThan(tree.symbol)
+        !tree.tpe.finalResultType.typeSymbol.primaryConstructor.isLessAccessibleThan(tree.symbol)
     }

     private def transformCaseApply(tree: Tree) = {
➜  sandbox git:(pr5002) qsc Test.scala
➜  sandbox git:(pr5002)
```

lrytz commented:
(Though that's not related to the fix of SI-9546, it's rather an improvement of #4675 that fixed the original issue SI-9425)

retronym committed:
SI-9425 Fix a residual bug with multi-param-list case classes

During code review for the fix for SI-9546, we found a corner
case in the SI-9425 that remained broken.

Using `finalResultType` peels off all the constructor param
lists, and solves that problem.

retronym commented:
> (Though that's not related to the fix of SI-9546, it's rather an improvement of #4675 that fixed the original issue SI-9425)

Good catch, I've pushed another commit with your test case that switches to `finalResultType`.

retronym commented:
@lrytz I've used this text:

```
Rewrite eligible calls to monomorphic case companion apply methods to the equivalent constructor call.

Note: for generic case classes the rewrite needs to be handled at the enclosing `TypeApply` to transform
`TypeApply(Select(C, apply), targs)` to `Select(New(C[targs]), <init>)`. In case such a `TypeApply`
was deemed ineligible for transformation (e.g. the case constructor was private), the refchecks transform
will recurse to this point with `Select(C, apply)`, which will have a type `[T](...)C[T]`.

We don't need to perform the check on the Select node, and `!isHigherKinded will guard against this
redundant (and previously buggy, SI-9456) consideration.
```

lrytz commented:
LGTM, thanks!

SethTisue commented:
community build of the merge commit: https://scala-ci.typesafe.com/job/scala-2.11.x-integrate-community-build/245/consoleFull (404 while Jenkins catches up)

SethTisue commented:
newer: https://scala-ci.typesafe.com/job/scala-2.11.x-integrate-community-build/246/consoleFull
